### PR TITLE
fix[codegen]: fix double evals in sqrt, slice, blueprint

### PR DESCRIPTION
**Pull request by cyberthirst**
### What I did
fix for:
- https://github.com/vyperlang/vyper/issues/3975
- https://github.com/vyperlang/vyper/issues/3972
- https://github.com/vyperlang/vyper/issues/3973
### How I did it
cache respective args
### How to verify it
added new tests for side-effect evaluations for the respective builtins
### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/builtins/functions.py [hunk 1]
```diff
@@ -1831,7 +1839,11 @@ def _build_create_IR(
             # return a complex expression which writes to memory and returns
             # the length of the encoded data
             argslen = abi_encode(argbuf, to_encode, context, bufsz=bufsz, returns_len=True)
+            return self._helper(
+                argbuf, bufsz, target, value, salt, argslen, code_offset, revert_on_failure
+            )
 
+    def _helper(self, argbuf, bufsz, target, value, salt, argslen, code_offset, revert_on_failure):
```
Comment (charles-cooper) [inline vyper/builtins/functions.py:1846]: not sure why this was added

Comment (charles-cooper) [reply vyper/builtins/functions.py:1846]: ah, missed a few lines up

File: tests/functional/builtins/codegen/test_slice.py [hunk 2]
```diff
@@ -536,3 +536,34 @@ def test_slice_buffer_oob_reverts(bad_code, get_contract, tx_failed):
     c = get_contract(bad_code)
     with tx_failed():
         c.do_slice()
+
+
+def test_slice_adhoc_locs_evals_once_side_effects(get_contract):
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_slice.py]: this test seems redundant with a few lines up

File: tests/functional/builtins/codegen/test_slice.py [hunk 3]
```diff
@@ -536,3 +536,34 @@ def test_slice_buffer_oob_reverts(bad_code, get_contract, tx_failed):
...
+    b: Bytes[32] = concat(convert(65, bytes32), b'')
+    c: Bytes[10] = slice(b, 31, 1)
+    return c
+    """
+
+    c = get_contract(code)
+    assert c.ret10_slice() == b"A"
+
+
+def test_sqrt_evals_once_side_effects(get_contract):
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_slice.py]: no need for the name to have both "eval once" and "side effects"

File: tests/functional/builtins/codegen/test_slice.py [hunk 4]
```diff
@@ -536,3 +536,22 @@ def test_slice_buffer_oob_reverts(bad_code, get_contract, tx_failed):
     c = get_contract(bad_code)
     with tx_failed():
         c.do_slice()
+
```
[the remark this answers is not in the document]
Comment (cyberthirst) [reply tests/functional/builtins/codegen/test_slice.py:539]: but i'll add the missing locations to make sure

File: tests/functional/builtins/codegen/test_slice.py (hunk 3 above)
Comment (cyberthirst) [reply tests/functional/builtins/codegen/test_slice.py]: renamed

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**